plugins/node/instrumentation-socket.io/src/socket.io.ts

```diff
@@ -0,0 +1,477 @@
+                  eventName,
+              },
+            }
+          );
+
+          if (self._config.onHook) {
+            safeExecuteInTheMiddle(
+              () =>
+                self._config?.onHook?.(span, { moduleVersion, payload: args }),
+              (e) => {
+                if (e) self._diag.error(`onHook error`, e);
+              },
+              true
+            );
+          }
+          return context.with(trace.setSpan(context.active(), span), () =>
+            self.endSpan(() => originalListener.apply(this, arguments), span)
+          );
+        };
+        return original.apply(this, [ev, wrappedListener]);
+      };
+    };
+  }
+
+  private endSpan(traced: () => any | Promise<any>, span: Span) {
+    try {
+      const result = traced();
+      if (isPromise(result)) {
+        return Promise.resolve(result)
+          .catch((err) => {
+            if (err) {
+              if (typeof err === "string") {
+                span.setStatus({ code: SpanStatusCode.ERROR, message: err });
+              } else {
+                span.recordException(err);
+                span.setStatus({
+                  code: SpanStatusCode.ERROR,
+                  message: err?.message,
+                });
+              }
+            }
+            throw err;
+          })
+          .finally(() => span.end());
+      } else {
+        span.end();
+        return result;
+      }
+    } catch (error: any) {
+      span.recordException(error);
+      span.setStatus({ code: SpanStatusCode.ERROR, message: error?.message });
+      span.end();
+      throw error;
+    }
+  }
+
+  private _patchEmit(moduleVersion: string) {
+    const self = this;
+    return (original: Function) => {
+      return function (this: any, ev: any, ...args: any[]) {
+        if (!self._config.traceReserved && reservedEvents.includes(ev)) {
+          return original.apply(this, arguments);
+        }
+        if (self._config?.emitIgnoreEventList?.includes(ev)) {
+          return original.apply(this, arguments);
+        }
+        const messagingSystem = "socket.io";
+        const eventName = ev;
+        const attributes: any = {
+          [SemanticAttributes.MESSAGING_SYSTEM]: messagingSystem,
+          [SemanticAttributes.MESSAGING_DESTINATION_KIND]:
+            MessagingDestinationKindValues.TOPIC,
+          [SocketIoInstrumentationAttributes.SOCKET_IO_EVENT_NAME]: eventName,
+        };
+
+        let rooms =
+          this.rooms ||
+          this._rooms ||
+          this.sockets?._rooms ||
+          this.sockets?.rooms ||
+          [];
+        // Some of the attributes above are of Set type. Convert it.
+        if (!Array.isArray(rooms)) {
+          rooms = Array.from<string>(rooms);
+        }
+        // only for v2: this.id is only set for v2. That's to mimic later versions which have this.id in the rooms Set.
+        if (rooms.length === 0 && this.id) {
+          rooms.push(this.id);
+        }
```

Comment:
nit: This is a great candidate to go into the utils.ts file if you want to.

Something like `extractRoomsAttributeValue(this)`